// redux-orm/lib/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});
exports.createSelector = exports.createReducer = exports.oneToOne = exports.attr = exports.many = exports.fk = exports.OneToOne = exports.ManyToMany = exports.ForeignKey = exports.Session = exports.Backend = exports.Schema = exports.ORM = exports.Model = exports.QuerySet = exports.Attribute = undefined;

var _QuerySet = require('./QuerySet');

var _QuerySet2 = _interopRequireDefault(_QuerySet);

var _Model = require('./Model');

var _Model2 = _interopRequireDefault(_Model);

var _ORM = require('./ORM');

var _Session = require('./Session');

var _Session2 = _interopRequireDefault(_Session);

var _redux = require('./redux');

var _fields = require('./fields');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

var Schema = _ORM.DeprecatedSchema;

var Backend = function RemovedBackend() {
    throw new Error('Having a custom Backend instance is now unsupported. ' + 'Documentation for database customization is upcoming, for now ' + 'please look at the db folder in the source.');
};

exports.Attribute = _fields.Attribute;
exports.QuerySet = _QuerySet2.default;
exports.Model = _Model2.default;
exports.ORM = _ORM.ORM;
exports.Schema = Schema;
exports.Backend = Backend;
exports.Session = _Session2.default;
exports.ForeignKey = _fields.ForeignKey;
exports.ManyToMany = _fields.ManyToMany;
exports.OneToOne = _fields.OneToOne;
exports.fk = _fields.fk;
exports.many = _fields.many;
exports.attr = _fields.attr;
exports.oneToOne = _fields.oneToOne;
exports.createReducer = _redux.createReducer;
exports.createSelector = _redux.createSelector;
exports.default = _Model2.default;
